DEVELOPERS.md

```diff
@@ -11,47 +11,109 @@ The `just` command provides a list of available recipes:
+```
+just test-unit --verbose --durations=10
+```
+
+For maximum flexibility, the `test` command can be used to run individual test files or tests, or to do other clever
+things with `pytest`. It just delegates to `pytest`. For example:
 ```
-just test-integration tests/acceptance
+just test tests/integration/backends/test_tpp.py
 ```
 
-To pass multiple args, wrap in quotes, e.g.:
+Since we have many tests that are parameterized across multiple databases and the Spark database tests are very slow,
+test commands which run such parameterized tests have a variant that skips the Spark tests. These are much faster and
+should be used unless you're specifically working on Spark. For example:
 ```
-just test-integration '-s tests/acceptance'
+just test-integration-no-spark
 ```
 
-To remove the persistent database containers:
+### Writing tests
+
+Please think carefully about how to test code that you are adding or changing. We think that test maintainability is a
+big risk for this system, so we're trying to be very deliberate about the kind of tests that we write. You should
+follow these guidelines and raise it with the rest of the team for discussion if you think that they are problematic.
+
+* _Major features in ehrQL or the Data Builder_ more widely should be motivated by a study used in an **acceptance**
+  test. The number of such studies should be kept small in order that they don't be come a maintenance burden. The
+  studies we use for the acceptance tests will need to be chosen carefully as representative of how we expect Data
+  Builder to be used; they may be real studies or synthetic ones as appropriate.
+* All _erhQL features_ should be covered by **spec** tests.
+* Complex _query language logic_ that is not fully covered by the spec tests should be covered by **unit** tests. To
+  avoid duplication, you should not write unit tests for logic that is adequately covered by the spec tests.
+* The main _functionality of query engines_ will naturally be covered by **spec** tests, which are run against all the
+  query engines.
+* Complex _query engine logic_ that is not fully covered by the spec tests should be covered by **unit** or
+  **integration** tests as appropriate. To avoid duplication, you should not write such tests for logic that is
+  adequately covered by the spec tests.
+* Functionality of the _Docker image_ and how it invokes Data Builder are covered by a small set of **docker** tests.
+* The adherence of _backends to contracts_ that they implement is automatically validated by **backend validation**
+  tests.
+* Where _backend tables_ do not map directly to the contracts that they implement, it may be helpful to write
+  **integration** tests.
+* All other _supporting logic_ should be covered by **unit** tests. Please avoid the temptation to cover this using
+  integration or docker tests that run the Data Builder end-to-end.
+* Do not write **legacy** tests. :-)
```

Comment:
Is this the place to discuss 100% coverage, and when it's acceptable to exclude lines from coverage?

Reply:
I think it would be good to add it alongside this content in some way, but probably not gate this PR on it?

Reply:
Too late!

Could one of you have a look at https://github.com/opensafely-core/databuilder/pull/410/commits/9d02b758d0dabb6334964cfcb7b25fa414c9a81a, please?